### PR TITLE
Feature/switch

### PR DESCRIPTION
**Pull request by CharlesNollet**
Add the switch view to trikot viewmodels

## Description
With this feature, a user can add a switch to his UI and bind it to a trikot viewmodel

## Motivation and Context
Adds versatility to apps that use trikot

## How Has This Been Tested?
- [x] All features has been added/updated in sample application ( /sample )
Integration has been tried in a current project with success

## Screenshots of sample application ( /sample ) (if appropriate):
Screen Shot 2021-02-02 at 10 14 43 AM

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: swift-extensions/UISwitchExtensions.swift [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            observe(switchViewModel.isOn) { [weak self] (value: Bool) in
+                self?.isOn = value
+            }
+        }
+    }
+
+    @objc
+    func onValueChanged(sender: UISwitch) {
+        if #available(iOS 10.0, *), let style = impactFeedbackStyle() {
+            UIImpactFeedbackGenerator(style: getFeedbackStyle(style: style)).impactOccurred()
```
Comment (marcantoinefortier) [inline swift-extensions/UISwitchExtensions.swift:28]: Nice 👍

File: viewmodels/src/commonMain/kotlin/com/mirego/trikot/viewmodels/SwitchViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+interface SwitchViewModel : ViewModel {
+    /**
+     * Currently selected state of the view
+     */
+    val isOn: Publisher<Boolean>
+
+    /**
+     * Set the state defined by the platform switch
+     */
+    fun changeState(state: Boolean)
```
Comment (marcantoinefortier) [inline viewmodels/src/commonMain/kotlin/com/mirego/trikot/viewmodels/SwitchViewModel.kt:14]:
```suggestion
    fun setIsOn(on: Boolean)
```

**Comment by CharlesMcD** [comment]
After doing it yesterday in our project I realized that we have to be careful with the term "switch". For exemple with the path "base/switch/SwitchViewModel" the generated code was throwing errors on android. We decided to go with the term "toggle" everywhere and to use the term "switch" only in the platform specific UI. 
Maybe something like `base/toggle/ToggleViewModel` or `base/toggleSwitch/ToggleSwitchViewModel` would be better.
```
@BindingAdapter("view_model", "lifecycleOwnerWrapper")
    fun bindSwitch(
        switch: SwitchCompat,
        viewModel: ToggleViewModel,
        lifecycleOwnerWrapper: LifecycleOwnerWrapper
    )
```
and 
```
extension UISwitch {
    var toggleViewModel: ToggleViewModel? {
        get { trikotViewModel() }
        set(value) {
            ......
        }
    }
}
```

File: android-ktx/src/main/java/com/mirego/trikot/viewModels/SwitchViewModelBinder.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+import com.mirego.trikot.viewmodels.mutable.MutableSwitchViewModel
+
+object SwitchViewModelBinder {
+    private val noSwitchViewModel =
+        MutableSwitchViewModel().apply { hidden = true.just() } as SwitchViewModel
+
+    @JvmStatic
+    @BindingAdapter("view_model", "lifecycleOwnerWrapper")
+    fun bind(
+        switch: SwitchMaterial,
```
Comment (CharlesMcD) [inline android-ktx/src/main/java/com/mirego/trikot/viewModels/SwitchViewModelBinder.kt:16]: `SwitchCompat` here for more flexibility ;)

File: android-ktx/build.gradle [hunk 4]
```diff
@@ -40,12 +40,13 @@ configurations.all {
 
 dependencies {
     implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
-    api 'com.mirego.trikot:viewmodels:1.0.6'
+    api 'com.mirego.trikot:viewmodels:1.0.11-SNAPSHOT'
```
Comment (marcantoinefortier) [inline android-ktx/build.gradle:43]: Je pense qu'il faut que tu le remette au latest ici?

```suggestion
    api 'com.mirego.trikot:viewmodels:1.0.10
```

File: sample/ios/.ruby-version [hunk 5]
```diff
@@ -1 +1 @@
-2.4.1
```
Comment (marcantoinefortier) [inline sample/ios/.ruby-version:1]: Nécessaire ce bump de ruby?

File: android-ktx/build.gradle (hunk 4 above)
Comment (CharlesMcD) [reply android-ktx/build.gradle:43]: The latest for android-ktx is `1.0.6`. 1.0.10 is the viewModel version

File: sample/ios/.ruby-version (hunk 5 above)
Comment (CharlesNollet) [reply sample/ios/.ruby-version:1]: bundle is only available on 2.7.2

File: android-ktx/build.gradle (hunk 4 above)
Comment (CharlesMcD) [reply android-ktx/build.gradle:43]: Ohh nevermind it's precisely the dependency to viewModel. So yeah I think we have to set 'com.mirego.trikot:viewmodels:1.0.11' but we have to do it after mirego-builds has done his `mirego.release pre tag commit` commit

File: swift-extensions/UISwitchExtensions.swift [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        }
+        switchViewModel?.changeState(state: self.isOn)
+    }
+
+    private func impactFeedbackStyle() -> FeedbackStyle? {
+        return objc_getAssociatedObject(self, AssosiatedKeys.impactFeedbackStyle) as? FeedbackStyle
+    }
+
+    private func impactFeedbackStyle(value: FeedbackStyle?) {
```
Comment (OlivierPineau) [inline swift-extensions/UISwitchExtensions.swift:37]: setImpactFeedbackStyle?

File: swift-extensions/UISwitchExtensions.swift [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+extension UISwitch {
+
+    private struct AssosiatedKeys {
+        static var impactFeedbackStyle = UnsafeMutablePointer<Int8>.allocate(capacity: 1)
+    }
+    
+    public var switchViewModel: SwitchViewModel? {
+        get { return trikotViewModel() }
+        set(value) {
+            if impactFeedbackStyle() == nil { impactFeedbackStyle(value: .light) }
```
Comment (OlivierPineau) [inline swift-extensions/UISwitchExtensions.swift:13]: We should probably not set a feedback style if it's nil

File: sample/android/src/main/java/com/trikot/viewModels/sample/databinding/ListItemViewModelAdapter.kt [hunk 8]
```diff
@@ -6,14 +6,7 @@ import com.mirego.trikot.viewmodels.adapter.GenericViewModelDiffCallback
...
 import com.trikot.viewmodels.sample.R
-import com.trikot.viewmodels.sample.viewmodels.ButtonListItemViewModel
-import com.trikot.viewmodels.sample.viewmodels.HeaderListItemViewModel
-import com.trikot.viewmodels.sample.viewmodels.ImageListItemViewModel
-import com.trikot.viewmodels.sample.viewmodels.InputTextListItemViewModel
-import com.trikot.viewmodels.sample.viewmodels.LabelListItemViewModel
-import com.trikot.viewmodels.sample.viewmodels.NavigableListItemViewModel
-import com.trikot.viewmodels.sample.viewmodels.SliderListItemViewModel
-import com.trikot.viewmodels.sample.viewmodels.ViewListItemViewModel
+import com.trikot.viewmodels.sample.viewmodels.*
```
Comment (marclefrancois) [inline sample/android/src/main/java/com/trikot/viewModels/sample/databinding/ListItemViewModelAdapter.kt:9]: Tes settings d'import ne sont pas bon, il ne faut pas mettre d'import `.*`

Comment (CharlesNollet) [reply sample/android/src/main/java/com/trikot/viewModels/sample/databinding/ListItemViewModelAdapter.kt:9]: ca le fait tout seul ... je vais fix ca

File: android-ktx/build.gradle (hunk 4 above)
Comment (CharlesNollet) [reply android-ktx/build.gradle:43]: doing so renders me unable to build locally and make sure everything works. how do i circumvent the problem?

Comment (marcantoinefortier) [reply android-ktx/build.gradle:43]: @npresseault could confirm, but I think our usual process is once we are sure every works locally, we simply "unsnapshot" our versions and merge our work. The next release build will then have the proper versions to work with.

File: swift-extensions/UISwitchExtensions.swift (hunk 6 above)
Comment (CharlesNollet) [reply swift-extensions/UISwitchExtensions.swift:37]: keeping the naming scheme used throughout the app

File: sample/android/src/main/java/com/trikot/viewModels/sample/databinding/ListItemViewModelAdapter.kt (hunk 8 above)
Comment (marcantoinefortier) [reply sample/android/src/main/java/com/trikot/viewModels/sample/databinding/ListItemViewModelAdapter.kt:9]: You still have a `viewmodels.*` on autoformat even when using the "Use single name import" Kotlin Code Style settings?
Screen Shot 2021-02-02 at 13 09 43

Comment (CharlesMcD) [reply sample/android/src/main/java/com/trikot/viewModels/sample/databinding/ListItemViewModelAdapter.kt:9]: Normalement tu devrais pouvoir empêcher un wildcard dans la config Editor > Code Style > Kotlin > Imports

**Pull request merged**